outlier detection: ensure ejection has priority over degradation

x48zhu opened the pull request:
### Commit Message:
Outlier detection supports degraded status. Hosts that return 2xx responses with the `x-envoy-degraded` header are marked as degraded by outlier detection. Degraded hosts are deprioritized in load balancing but remain in rotation (not ejected).

### Additional Description:
[Degrade]((https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/load_balancing/degraded)) is an existing feature for active health check. Outlier detection (a.k.a. passive health check) doesn't recognize degraded now. This change adds degraded health status to the outlier detection.

Key implementations:
- Degrade/undegrade is implemented mirrors existing eject/uneject [logic](https://www.envoyproxy.io/docs/envoy/v1.36.2/intro/arch_overview/upstream/outlier.html#ejection-algorithm)
	- Same backoff algorithm: Backoff increments on every degradation, and hence decrements after host stays healthy for one interval. Degrade reuses the same eject config parameters (e.g. base_ejection_time, max_ejection_time)
	- The only difference is, a successful active health check doesn't undegrade a host. A degrade host is only undegraded after the period of time.
- Degrade/undegrade uses has its own code, and is non-intrusive to the eject/uneject code. It updates a new host health flag `DEGRADED_OUTLIER_DETECTION`.
- Degraded responses count as successful for success rate calculations.

### Risk Level: Medium
This is a new optional feature that is disabled by default. When enabled, it changes load balancing behavior by deprioritizing hosts that report degraded state.
The implementation of degrade/undegrade mirrors the existing eject/uneject, but separate code, so it's non-intrusive to eject/uneject logic. 

### Testing
- Unit tests added in outlier_detection_impl_test.cc covering:
    - Degraded host marking via x-envoy-degraded header
    - Timer-based recovery with exponential backoff
    - Priority of ejection over degradation for 5xx responses
- Tests verify integration between router and outlier detection components with local build and triggering degrade via outlier detection.

### Docs Changes
- Added new configuration field `detect_degraded` in api/envoy/config/cluster/v3/outlier_detection.proto
- Added new host status `failed_outlier_degraded_check` in api/envoy/admin/v3/clusters.proto 

### Release Notes
Updated changelog/current.yaml

### Platform Specific Features
N/A

### Runtime guard
The feature is guarded by outlier detection configs. 

### Fixes #Issue: 
https://github.com/envoyproxy/envoy/issues/40053

abeyad commented:
/assign @cpakulski

adisuissa commented:
@tonya11en can you please take a first pass?
/assign @tonya11en

tonya11en reviewed (changes requested):
This is very useful, thanks! There should be a new metric to let us monitor ejections caused by degradation. The rest of the comments are minor.

tonya11en commented:
/wait

x48zhu commented:
@tonya11en Thank you for the review! I have made the change according to your comments.
I haven't merged main branch. I will do so and resolve conflict before merging the PR.
I believe [Envoy outlier detection doc](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/outlier) also needs an update. Could you please suggest how to do so if needed?

x48zhu commented:
@cpakulski Thank you for the comment! I have changed the specific code pointed by your comment to reuse `putHttpResponseCode`.
There are other duplication, such as the new function `checkHostForUndegrade` mimics existing function `checkHostForUneject`, where the common logic is possibly extracted and reused. However I personally think it adds complexity and reduce reability. I also would like degrade logic to be least intrusive to the existing logic. If you prefer to remove any other duplication, please let me know. I'm happy to make further changes. Thank you!

phlax commented:
@x48zhu this needs a main merge

@cpakulski i think this is waiting on further review

x48zhu commented:
/retest

x48zhu commented:
Hi @cpakulski  @phlax @tonya11en could you please help me proceed with the PR review? Thank you!
(I merged the master but one CI test is failed with error that's irrelevant to this PR. I think a force push may help but didn't do so . Could you please suggest the best action)

cpakulski commented:
/retest

ravenblackx commented:
Pinged @tonya11en and @abeyad

tonya11en reviewed (approved):
Sorry for the delay, folks.

x48zhu commented:
/retest

x48zhu commented:
Hi @tonya11en I made the change by adding jitter to the undegrade window.
I also noticed that there's a [ejection event logging](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/outlier#ejection-event-logging) which I didn't include degrade event, which I implement now in the latest change.
Please take a look. Thank you!

tonya11en commented on api/envoy/data/cluster/v2alpha/outlier_detection_event.proto:
```diff
@@ -64,6 +64,12 @@ enum OutlierEjectionType {
+  // The host returns responses with the x-envoy-degraded header, indicating it is under stress
+  // but still able to serve traffic. Degraded hosts are deprioritized in load balancing but not
+  // fully ejected.
+  DEGRADED = 7;
```
This is not needed. Just the v3

wbpcode reviewed (commented):
Thanks for this contribution. And one high level comment to start this review.

/wait-any

x48zhu commented on source/common/upstream/outlier_detection_impl.cc:
```diff
@@ -373,6 +404,31 @@ void DetectorImpl::checkHostForUneject(HostSharedPtr host, DetectorHostMonitorIm
+      event_logger_->logUneject(host);
+    }
+  }
+}
```
Thank you for reviewing! I'm proposing a timer-based as:
1. It's consistent with how active health checking marks a host as degraded when the header is present and clear it when the header is absent.
2. The timer-based recovery aligns with how other outlier detection types work (e.g. consecutive 5xx). I believe the reason is, once a host is ejected or degraded, it's removed from normal load balancing rotation and stops receiving traffic. If we relied on checking response headers to bring the host back, the host can possbily never respond with headers to signal it has recovered. The timer gives host another opportunity to receive traffic and demonstrate it's healthy again.

I hope this makes sense. I'm open to adjusting the design.

wbpcode commented:
@x48zhu Thanks for this contribution. And I also checked the original requirement. I have a question before to take a deep check to the source code. (Because the complexity of current outlier detection, if want to avoid this change if possible). NOTE: degrade host is not unhealthy host and will still take traffic if there are no enough healthy host.

Could this https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/upstreams/http/v3/http_protocol_options.proto#envoy-v3-api-msg-extensions-upstreams-http-v3-httpprotocoloptions-outlierdetection meet your requirement? With this feature, we can mark an response with specific header as error? Then the host will be ejected? which should could meet your requirement in the https://github.com/envoyproxy/envoy/issues/40053.

For outlier detection, I don't think we need to distinguish a new status of degraded. I think we only need to support:
OK: this host could accept traffic.
NO: this host is cannot accept traffic for some reason, error, degraded and so on.

(I have same opinion to active health checker, but it have been supported there...)

The https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/upstreams/http/v3/http_protocol_options.proto#envoy-v3-api-msg-extensions-upstreams-http-v3-httpprotocoloptions-outlierdetection is designed to handle these special cases because we don't want to make the outlier detection more and more complex.

wbpcode commented on source/common/upstream/outlier_detection_impl.cc:
```diff
@@ -571,6 +637,91 @@ void DetectorImpl::onConsecutiveLocalOriginFailure(HostSharedPtr host) {
+    return;
+  }
+  notifyMainThreadHostUndegraded(host);
+}
```
When will this be called?

x48zhu replied:
These are not used so I removed them.

wbpcode commented on source/common/router/router.cc:
```diff
@@ -1740,14 +1740,24 @@ void Filter::onUpstreamHeaders(uint64_t response_code, Http::ResponseHeaderMapPt
+  } else {
+    upstream_request.upstreamHost()->outlierDetector().putResult(
+        Upstream::Outlier::Result::ExtOriginRequestSuccess, response_code_for_outlier_detection);
+  }
```
Please add a test to cover these change. Thanks.

wbpcode reviewed (commented):
Thanks for the contribution. The code change LGTM overall. And only some minor comments are added.

/wait

x48zhu commented:
/retest

x48zhu commented:
@wbpcode Thanks for the review! I addressed your comment in the [last commit](https://github.com/envoyproxy/envoy/pull/41852/changes/db05ef5ac60c199d1d6c74e4e353d1acbb7b71bf). 
(I accidentally force push, but I don't amend any previous commit. I hope this doesn't cause any review trouble. 🙇 )
